### PR TITLE
[SPARK-33388][SQL] Merge In and InSet predicate

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
According to https://github.com/apache/spark/pull/28269#issuecomment-655365714
Create a basic class for `In` and `InSet`.

### Why are the changes needed?
No

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existed UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -561,14 +548,84 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
...
+  }
+
+  override def toString: String = s"$value IN ${list.mkString("(", ",", ")")}"
+}
+
+/**
+ * Optimized version of In clause, when all filter values of In clause are
+ * static.
+ */
+case class InSet(value: Expression, list: Seq[Expression]) extends BasicIn {
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:603]: Could we avoid change `Set[Any]` to `Seq[Expression]`?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:603]: > Could we avoid change `Set[Any]` to `Seq[Expression]`?

Yea, it seems achievable， updated

**Comment by AngersZhuuuu** [comment]
gentle ping @cloud-fan

**Comment by dongjoon-hyun** [comment]
I guess this needs @cloud-fan 's approval.

**Comment by AngersZhuuuu** [comment]
> I guess this needs @cloud-fan 's approval since this is his request.

Yea, wait for his review

**Comment by ulysses-you** [comment]
I think that the merge is not only code side. We have many code place use the `In` and `InSet`, if we decide to merge them then the origin idea should like that how to promise wherever we use and whatever use `In` or `Inset` or `InBase` we can have the best performance.

**Comment by AngersZhuuuu** [comment]
> I think that the merge is not only code side. We have many code place use the `In` and `InSet`, if we decide to merge them then the origin idea should like that how to promise wherever we use and whatever use `In` or `Inset` or `InBase` we can have the best performance.

Good suggestion, in this way we can avoid complex convert code and do all the thing in `InBase`

gentle ping @cloud-fan Hope for your suggestion.

**Pull request closed without merging**